cdap-docs/developers-manual/source/building-blocks/applications.rst

```diff
@@ -48,9 +48,78 @@ Other components are defined using user-written classes that implement correspon
 interfaces and are referenced by passing an object, in addition to being assigned a unique
 name.
 
-Names used for streams and datasets need to be unique across the CDAP instance, while
+Names used for streams and datasets need to be unique across the CDAP namespace, while
 names used for programs and services need to be unique only to the application.
 
+Applications can use a ``Config`` class to receive a configuration during deployment time of the Application.
+For example, this can be used to specify the stream, dataset to create/read from at application deployment time
+instead of hard-coding them in the ``AbstractApplication``'s ``configure`` method. The configuration class needs to be
+the type parameter of the ``AbstractApplication`` class. It should also extend the ``Config`` class present in the CDAP API.
+The configuration is provided during application deployment time through a Header in the RESTful API. It is available during
+configuration time through the ``getConfig()`` method in ``AbstractApplication``.
+
+Information about the RESTful call is available in the :ref:`Lifecycle HTTP RESTful API documentation <http-restful-api-lifecycle>`.
+
+We can modify the ``MyApp`` class above to take in a Configuration ``MyApp.MyAppConfig``::
+
+      public class MyApp extends AbstractApplication<MyApp.MyAppConfig> {
+
+        public static class MyAppConfig extends Config {
+          String streamName;
+          String datasetName;
+
+          public MyAppConfig() {
+            // Default values
+            this.streamName = "myAppStream";
+            this.datasetName = "myAppDataset";
+          }
+        }
+
+        @Override
+        public void configure() {
+          MyAppConfig config = getConfig();
+          setName("myApp");
+          setDescription("My Sample Application");
+          addStream(new Stream(config.streamName));
+          createDataset(config.datasetName, Table.class);
+          addFlow(new MyAppFlow(config));
+          addService(new MyService(config.datasetName));
+          addMapReduce(new MyMapReduce(config.datasetName));
+          addWorkflow(new MyAppWorkflow());
+        }
+      }
+
+In order to use the configuration in programs, we pass it to individual Programs using their constructor. If
+the configuration parameter is also required during runtime, you can use the ``@Property`` annotation::
```

Comment:
`If the configuration parameter is also required during runtime` is a bit confusing. It is not clear when the property passed through constructor argument is used vs property with `@Property` annotation.